### PR TITLE
Feature/sbachmei/mic 4236 copy changelog to make results dir

### PR DESCRIPTION
**Pull request by stevebachmeier**
### Description

- *Category*: post-processing
- *JIRA issue*: [MIC-4236](https://jira.ihme.washington.edu/browse/MIC-4236), [MIC-3940](https://jira.ihme.washington.edu/browse/MIC-3940)
- *Research reference*: na

### Changes and notes
 
The main goal here is to copy the newly-created CHANGELOG.rst file to the
final results directory to help folks understand what to look for when 
V&Ving results.

While at it, a few other minor changes:
- Bugfix symlink generation so that they're only created at the end
    (was an issue w/ jobmon -> MIC-3940)
- minor refactor to move results directory helper functions into new cli_utilities.py
- move version validation into cli_utilities.py
- Bugfix version validation
- Custom `handle_exceptions` function. Unlike vivarium's utility function previously
  used, this `raise`s even with --pdb. The original functionality didn't and that would
  result in marking bad results as latest and best if you used `c` to "continue" out of
  pdb instead of ctrl+c. We might want this to be standard behaviour of `handle_exceptions`?

NOTE: this moves the symlinks to the `.../final_results/` directory. This is different than before

image

### Verification and Testing

Ran a short sim and validated no regression and CHANGELOG exists in 
final results directory.

### CODE REVIEW

File: src/vivarium_census_prl_synth_pop/tools/cli.py [hunk 1]
```diff
@@ -178,16 +179,8 @@ def make_results(
 ) -> None:
     """Create final results datasets from the raw results output by observers"""
     configure_logging_to_terminal(verbose)
+    validate_args(mark_best=mark_best, test_run=test_run, label_version=label_version)
     logger.info("Creating final results directory.")
-    if label_version is not None:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/cli.py:182]: I moved this into `validate_args`

File: src/vivarium_census_prl_synth_pop/tools/cli_utils.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+def validate_args(
+    mark_best: bool, test_run: bool, label_version: Optional[str] = None
+) -> None:
+    if mark_best and test_run:
+        raise RuntimeError(
+            "A test run can't be marked best. "
+            "Please remove either the mark best or the test run flag."
+        )
+    if label_version is not None:
+        expected_version_format = re.compile("^\d+\.\d+\.\d+$")
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/cli_utils.py:22]: Bugfix here: the original regex was allowing for non-digit parts (foo.bar.baz) and also was not limiting it to only three (1.2.3.4.5.6)

File: src/vivarium_census_prl_synth_pop/tools/cli_utils.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        if expected_version_format.match(label_version):
+            pass
+        else:
+            raise ValueError(
+                f"Version '{label_version}' is not of correct format. "
+                "Format for version should be '#.#.#'"
+            )
+
+
+def build_final_results_directory(
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/cli_utils.py:32]: This is a direct copy from utilities.py to here

File: src/vivarium_census_prl_synth_pop/tools/cli_utils.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    final_output_dir = build_output_dir(
+        Path(results_dir),
+        subdir=subdir,
+    )
+    raw_output_dir = Path(results_dir) / paths.RAW_RESULTS_DIR_NAME
+
+    return raw_output_dir, final_output_dir
+
+
+def finish_post_processing(final_output_dir: Path, test_run: bool, mark_best: bool) -> None:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/cli_utils.py:50]: this method is totally new - please review

File: src/vivarium_census_prl_synth_pop/tools/make_results.py [hunk 5]
```diff
@@ -246,52 +246,27 @@
...
+    raw_output_dir: Path,
+    final_output_dir: Path,
     mark_best: bool,
     test_run: bool,
     extension: str,
     public_sample: bool,
     seed: str,
-    artifact_path: Union[str, Path],
-):
-    if mark_best and test_run:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/make_results.py:258]: moved this into cli_utils.validate_args method

File: src/vivarium_census_prl_synth_pop/tools/make_results.py [hunk 6]
```diff
@@ -246,52 +246,27 @@
...
-    final_output_dir = Path(final_output_dir)
-    artifact_path = Path(artifact_path)
+    artifact_path: str,
+) -> None:
     logger.info("Performing post-processing")
     perform_post_processing(
         raw_output_dir, final_output_dir, extension, seed, artifact_path, public_sample
     )
 
-    if test_run:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/make_results.py:272]: moved this into cli_utils.validate_args method

File: src/vivarium_census_prl_synth_pop/tools/make_results.py [hunk 7]
```diff
@@ -246,52 +246,27 @@
...
-        logger.info("Test run - not marking results as latest.")
-    else:
-        create_results_link(final_output_dir, paths.LATEST_DIR_NAME)
-
-    if mark_best:
-        create_results_link(final_output_dir, paths.BEST_DIR_NAME)
-
-
-def create_results_link(output_dir: Path, link_name: Path) -> None:
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/make_results.py:281]: moved this into cli_utils.finish_postprocessing

File: src/vivarium_census_prl_synth_pop/utilities.py [hunk 8]
```diff
@@ -288,24 +287,6 @@ def build_output_dir(output_dir: Path, subdir: Optional[Union[str, Path]] = None
     return output_dir
 
 
-def build_final_results_directory(
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/utilities.py:291]: moved this into cli_utils.py

File: src/vivarium_census_prl_synth_pop/tools/cli_utils.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def _create_results_link(output_dir: Path, link_name: Path) -> None:
+    logger.info(f"Marking results as '{link_name}': {str(output_dir)}.")
+    # Create simulation directory link (two levels higher than output_dir)
+    # Create relative link
+
+    # output_root_dir = output_dir.parent
+    # link_dir = output_root_dir / link_name
+    link_dir = (
```
Comment (stevebachmeier) [inline src/vivarium_census_prl_synth_pop/tools/cli_utils.py:77]: This symlink location has changed from before - it's now plopped a few levels higher right in the `.../final_results/` directory

File: src/vivarium_census_prl_synth_pop/tools/cli_utils.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        Path(str(output_dir).split(str(paths.FINAL_RESULTS_DIR_NAME))[0])
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/tools/cli_utils.py:102]: You should be able to use use `path.parent` to navigate up the directory tree.

File: src/vivarium_census_prl_synth_pop/tools/cli.py [hunk 11]
```diff
@@ -57,7 +58,7 @@ def make_artifacts(
     with_debugger: bool,
 ) -> None:
     configure_logging_to_terminal(verbose)
-    main = handle_exceptions(build_artifacts, logger, with_debugger=with_debugger)
+    main = handle_exceptions(build_artifacts, with_debugger=with_debugger)
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/tools/cli.py:61]: What happened to the logger? Why don't we need it here anymore?

File: src/vivarium_census_prl_synth_pop/tools/cli.py [hunk 12]
```diff
@@ -178,19 +179,11 @@ def make_results(
 ) -> None:
     """Create final results datasets from the raw results output by observers"""
     configure_logging_to_terminal(verbose)
-    logger.info("Creating final results directory.")
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/tools/cli.py:181]: Any reason you removed this log message?

**Review by rmudambi — APPROVED** [review]
Thanks for the comments about what had changed versus just having been moved.

**Pull request merged**